### PR TITLE
ruby: introduce Gherkin::Query#parent_locations.

### PR DESCRIPTION
**Pull request by botandrose**
### 🤔 What's changed?
Introduces a new Gherkin::Query#parent_locations method in the Ruby implementation.

Cucumber::Core::Compiler will use this method to tell Cucumber::Core::Test::Case the locations of its parent lines, such as `Feature:`, `Background:`, and `Rule:`, so that it will correctly match them during location filtering. This particular strategy was identified as potentially reasonable while pairing with @mattwynne.

### ⚡️ What's your motivation? 

This is one part of a two-repo PR to resolve the last remaining bits of the regression described in https://github.com/cucumber/cucumber-ruby/issues/1469

### 🏷️ What kind of change is this?

- :zap: New feature (non-breaking change which adds new behaviour)

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [x] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [?] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [?] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

I'm not sure about these last two, since this could be considered as internals from a high-level cucumber point-of-view, but a public-interface addition from the gem point-of-view. Thoughts?

### CODE REVIEW

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [inline ruby/lib/gherkin/query.rb:32]: 5second check. Should this be `child&rule` to cover where the child does / doesn't respond to rule?

Comment (luke-hill) [reply ruby/lib/gherkin/query.rb:32]: Right, so this should be added then right?

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:32]: Hmm, I'm not following. Are you saying that we might want to change line 30 by adding a safe navigation operator (&.) to the postfix guard clause, like so?

`update_rule(feature, child.rule) if child&.rule`

Comment (luke-hill) [reply ruby/lib/gherkin/query.rb:32]: No the line I've commented on here - 32. you have a `nil` rule. this feels wrong?

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:32]: @luke-hill Ahh, I gotcha now. Let me dig in and see if I can figure out why I thought `nil` made sense back when I wrote this.

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:32]: @luke-hill I really can't figure out why I put `nil` there. All the tests still pass when replacing it with `child.rule`. No need for safe nav operator I don't think, since child always responds to `#rule` (it just might be nil, which is handled in the called method). I'll push a commit.

Comment (botandrose) [reply ruby/lib/gherkin/query.rb]: @luke-hill Okay, pushed. Could probably do with a squash when merging.

**Comment by luke-hill** [comment]
So I've had another quick scan of this @botandrose - I think firstly this probably "feels" like a major change, because you're refactoring a huge chunk of how features e.t.c. are updated (I know they're private), but this entire area is a private DSL essentially.

Looking at the code changes (Not spec), I actually think it all makes sense and is fine. Next step would be to review specs.

**Comment by luke-hill** [comment]
@botandrose so I'm keeping you in the loop and working "semi" linearly. Are we saying this should / must go in first and be released before thinking about other fixes?

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [inline ruby/lib/gherkin/query.rb:45]: Think this should be `rule` based on how you've done it in the other methods.

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [inline ruby/lib/gherkin/query.rb:47]: Then I'm assuming this might need tweaking?

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [inline ruby/lib/gherkin/query.rb:60]: To save time, as this is the only bit of the new API that's completely new, can you explain it in a line or two of prose so I don't need to dig into the whole API surface for gherkin

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [inline ruby/lib/gherkin/query.rb:62]: can the safe navigation operators here ever return nils i.e. be needed?

**Review by luke-hill — COMMENTED** [review]
I think by and large this might be the last review I do on this, and just go off a bit of faith as it's still a bit alien to me.

**Comment by botandrose** [comment]
@luke-hill Yes, my cucumber-ruby-core PR is built on-top of this new API, so this has to land before that one can take advantage of it. And they are essentially a pair... without one, the other doesn't really make any sense.

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:45]: This is the more abstract `parent`, because sometimes the direct parent of a Background is a Feature, and other times its a Rule. For example:
```
Feature: Cool feature is cool
Scenario: I become cool by using cool feature
```

```
Feature: Cool feature is cool
Rule: Only uncool people can become cool
Scenario: Already cool person cannot use cool feature
```

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:47]: IMO, the "correct" thing to do here would be to add a `Cucumber::Messages::Feature#id` method, to make it consistent with the rest of the message types, and the we could get rid of this workaround. It seems like an obvious hole to me. But it also seemed out of scope for this PR.

It seems to me that we have some options:
1. I can push another commit adding `C::M::F#id`
2. I can do a follow-up PR cleaning this up

I'm leaning towards 2 to avoid scope creep, but you're the boss! What do you think?

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:60]: Sure, this basically is the meat of the entire effort to fix the long-standing Cucumber 4.0 regression https://github.com/cucumber/cucumber-ruby/issues/1469

This block of code is collating all the line numbers of all the parents of the given scenario:
1. The Feature (and its optional description lines)
2. The optional Background and its steps for the _Scenario_
3. The optional Rule line
4. The optional Background and its steps for the _Rule_

It stores these line numbers data structure -- keyed to the scenario -- for later lookup.

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:62]: Yeah, the first item is required, but the latter three are optional. In other words, a Scenario always belongs to a Feature (line 1), but may or may not belong to a Background (line 2), a Rule (line 3), and/or a Rule Background (line 4), so we want to handle the nil case for the last three lines.

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:47]: @luke-hill Oh derp, 1 is impossible because its a different repo.

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:47]: @luke-hill I guess option 1 would be opening a PR on https://github.com/cucumber/messages , and waiting for that to be merged and released, and then we could merge and release this with a bumped requirement on that version, and then finally merge and release the cucumber-ruby-core PR with a bumped requirement on this.

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:47]: @luke-hill Just checked with crossed fingers, but `C::M::Feature` is still missing `#id` on current master : https://github.com/cucumber/messages/blob/main/ruby/lib/cucumber/messages.deserializers.rb#L244

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:47]: @luke-hill Ugh, and it looks like that code is all auto-generated. Adding an id to feature would change the data structure for ALL implementations of cucumber, not just the ruby one.

Maybe the simplest method would be to just use `#object_id` as the internal data structure key for both Feature and Rule. Or maybe just the object itself!

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:62]: Again, as a curiosity, you can see that we're keying the rule by its `#id`, but we have to use `#object_id` for the feature. :(

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb]: @luke-hill Okay, I just pushed a commit that uses the object itself as the key instead, which ends up simplifying everything greatly, including removing the need for so many safe navigation operators. I'm happy with this if you are!

File: ruby/lib/gherkin/query.rb
Comment (luke-hill) [reply ruby/lib/gherkin/query.rb:59]: Could you dump an example structure. I'm anticipating something like (With dodgy linespacing delib)

```
1 - Feature: Foo
2 - 
3 - Desc of feature
4 - 
5 - Background:
6 -  Given background one
7 - Scenario: Bar
8 - Given baz
9 - Then foo
```

Produces

```
{ feature: [1,3], background: [5], Rule: [], RuleBackground: []}
```

Comment (luke-hill) [reply ruby/lib/gherkin/query.rb:59]: And for my benefit. We're only looking at **parents** here because `locations` returns everything from the scenario down right?

Comment (botandrose) [reply ruby/lib/gherkin/query.rb:59]: @luke-hill Sure, I just changed the `expect` to `puts(...).inspect` in `spec/gherkin/query_spec.rb#L101`, which tests the method with the "Scenario: a ruled scenario" scenario and this is the result (I added line numbers here for easier output interpretting):

```gherkin
1
2    @feature-tag
3    Feature: my feature
4
5      Background:
6        Given a passed background step
7
8      @scenario-tag
9      Scenario: my scenario
10        Given a passed step
11
12      Scenario Outline: with examples
13        Given a <Status> step
14
15        @examples-tag
16        Examples:
17          | Status |
18          | passed |
19          | failed |
20
21      @rule-tag
22      Rule: this is a rule
23        Background:
24          Given the passed step in the rule background
25
26        @ruled-scenario-tag
27        Scenario: a ruled scenario
28          Given a step in the ruled scenario
```

```ruby
[#<Cucumber::Messages::Location:0x00007f4b2805a620 @line=3, @column=5>,
#<Cucumber::Messages::Location:0x00007f4b28158ea0 @line=5, @column=7>,
#<Cucumber::Messages::Location:0x00007f4b2805bd68 @line=22, @column=7>,
#<Cucumber::Messages::Location:0x00007f4b2804efc8 @line=23, @column=9>]
```

To answer your second question, you might notice that some lines are missing, like tag lines (2, 21, 26) and step lines (6, 24, 28). These lines are already matched correctly in `Cucumber::Core::Compiler#source_lines_for_pickle`. The companion PR over there adds a `#parent_source_lines_for_pickle` that uses this PR's `Gherkin::Query#scenario_parent_locations` method to match the last remaining missing pieces, the Feature, the Background, the Rule, and the Rule Background` lines, which is what you see in the output here.

I first looked at extending `#source_lines_for_pickle` to return everything, but other parts of the cucumber internals were relying on only the source lines being returned, not these extra parent lines, so we needed to keep them separate in the `C::C::TestCase` data structure. Hence, this additional method.

Comment (luke-hill) [reply ruby/lib/gherkin/query.rb:59]: Absolutely perfectly explained.

I personally dislike pos array returning bits over a hash. But I'm 99.9% certain you're going to say the other methods work the same way. i.e. it's magic that the first item is the feature ref

**Comment by luke-hill** [comment]
Approved CI to run. Think we're definitely 90% of the way there now. Sorry for the delay, I'm coming back after a long time off.

File: ruby/lib/gherkin/query.rb
Comment (botandrose) [reply ruby/lib/gherkin/query.rb:59]: I agree with you on that one, but the good news is that we don't actually care about where the lines come from, only that they match the given scenario. I.e. we could append a #shuffle to the array before returning it, and it wouldn't affect a thing.

**Comment by botandrose** [comment]
@luke-hill all good! I really appreciate your time on this and my other contributions throughout the years. Cucumber is my most used testing framework, as well, so again, your time and effort are very much appreciated

**Review by luke-hill — COMMENTED** [review]
Can you add a changelog entry for this, we'll look into getting this approved and merged shortly.

**Comment by botandrose** [comment]
@luke-hill Okay, added a changelog entry, rebased on current main, squashed, force-pushed. I think this is ready for merge!

**Review by luke-hill — APPROVED** [review]
Sorry for delays.

**Comment by aslakhellesoy** [comment]
Hi @botandrose,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by luke-hill** [comment]
This will go into the next major (v27), eta a week or so.

**Comment by botandrose** [comment]
@luke-hill @mattwynne thanks to both of you for your help on this! When v27 is out the door I'll rebase the other PR.

**Pull request merged**
